README.md
# ANSI escape sequence parser (also known as ECMA-48 / ANSI X3.64 / ISO/IEC 6429)

## Read this first

The first thing to understand when it comes to terminals is that [terminfo](https://en.wikipedia.org/wiki/Terminfo) (and [termcaps](https://en.wikipedia.org/wiki/Termcap)) exist.

The second thing to understand is that they exist for a reason. Various terminals and terminal emulators come with their own set of escape sequences. For example, `curses`-based software typically uses terminfo (or `termcap`) to output the correct codes for whatever terminal environment it is running in.

This package basically just plows ahead and directly parses things which are directly defined in the main ECMA-48 standard. Never-the-less, it may be helpful to understand that this is not the whole story.

## What is this useful for?

My impetus for creating this package is to use it in automating terminal software (interacting via a pseudo-terminal; see [posix-pty](http://hackage.haskell.org/package/posix-pty)).

The monad transformer `ParsecT e s m`, provided by megaparsec, allows one to mix (caveat emptor) application state, logic and even effects directly into the parser.

This also allows one to branch, altering the future behaviour of the parser based on past parse results. For example, you may care about this if you would like to support some kind of mode switching (e.g. ISO/IEC 2022).

## The landscape

The history of terminal escape sequences is super confusing (at least, to me). Standards are called by various names; have various revisions; they overlap and interact in non-trivial ways; they have subtly different raisons d'être... and so on.

I'll make no attempt to go into historical details, but it is helpful to at least have a rough sort of mental map of how
the standards relate to each other in order to orient you.

What follows is my best attempt at providing a little orientation and should not be taken as authoritative.
In fact, it is probably wrong on several points. (I personally have no great passion for arcane control codes.
Never-the-less, I am grateful to anyone with patience to find and point out inaccuracies.)

### Standards

* [**Unicode**](https://en.wikipedia.org/wiki/ASCII#Character_set)
* [**ASCII**](https://en.wikipedia.org/wiki/ASCII#Character_set)
  * A 7-bit code
  * Includes graphical as well as control characters (but no sequences)
  * Revisions:
    * ASA X3.4-1963
    * ASA X3.4-1965
    * USAS X3.4-1967
    * USAS X3.4-1968
    * ANSI X3.4-1977
    * ANSI X3.4-1986
    * ANSI X3.4-1986 (R1992)
    * ANSI X3.4-1986 (R1997)
    * ANSI INCITS 4-1986 (R2002)
    * ANSI INCITS 4-1986 (R2007)
    * ANSI INCITS 4-1986 (R2012)
  * Also known as: ECMA-6, ISO/IEC 646, and ITU T.50
* UTF-7
  * An encoding of unicode text using a (somewhat human readable) stream of 7-bit ASCII characters
* Extended ASCII
  * An 8-bit code
  * Sometimes confused with true ASCII
  * Adds additional characters in the range 128 to 255
  * Multiple codepages with different characters in the extended region
    * Typical code pages: ([1](https://en.wikipedia.org/wiki/Extended_ASCII#/media/File:Table_ascii_extended.png), [2]())
  * Also known as: ISO/IEC 8859, ECMA-94
* [**UTF-8**](https://en.wikipedia.org/wiki/UTF-8#Codepage_layout)
  * An 8-bit code
  * Modern and popular encoding of unicode text
  * Fully backward compatible with 7-bit ASCII
  * Somewhat backward compatible with Extended ASCII via either fall back or translation
  * Used in many 
* **ECMA-48**
  * A 7-bit code
  * Compatible with 7-bit ASCII
  * Also known as: ISO/IEC 6429
  * An evolution from ECMA-6, a ratified form of ISO/IEC 646.
* ISO/IEC 2022
  * An encoding for switching between character sets
  * Equivalent to ECMA-35. According to "someone on the internet":
    * ECMA-35, 1st Edition (1971) apparently described the character set facilities of the VT100 exactly, but was lost (!).
    * ECMA-35, 2nd edition, closer to VT220
  * Has disadvantages compared to UTF-8
* ANSI X3.64
  * Appears to have combined (earlier versions perhaps?) of ECMA-48 and ECMA-35 into one standard (?)
  * Finally withdrawn and replaced by ISO/IEC 6429 (equivalent to ECMA-48 5th edition)

    ¯\_(ツ)_/¯

Honestly. I don't know. The standards to probably pay attention to has been bolded.
That is to say, everything else is probably safe to ignore and included only for context.

### Terminology

* [Terminal](https://en.wikipedia.org/wiki/Computer_terminal)
* [Terminal emulator](https://en.wikipedia.org/wiki/Terminal_emulator)
* [Pseudoterminal](https://en.wikipedia.org/wiki/Pseudoterminal)
* [Character encoding](https://en.wikipedia.org/wiki/Character_encoding)
* [Code point](https://en.wikipedia.org/wiki/Code_point)
* [Code page](https://en.wikipedia.org/wiki/Code_page)
* **Control function**
* **Control character**
* **Control sequence**
* **Escape sequence**
* **Control string**

## Coverage

| ISO-2022

## Reference

* [Linux console codes](https://linux.die.net/man/4/console_codes)



ChangeLog.md
# Revision history for autoterm

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Main.hs
module Main (main) where

import Test.Hspec
import Test.Hspec.Runner

import qualified Text.Megaparsec.ANSISpec
import qualified Text.Megaparsec.ANSI.InternalSpec
import qualified Text.Megaparsec.ANSI.LexerSpec

spec :: Spec
spec = do
  Text.Megaparsec.ANSISpec.spec
  Text.Megaparsec.ANSI.InternalSpec.spec
  Text.Megaparsec.ANSI.LexerSpec.spec

main :: IO ()
main = hspecWith defaultConfig spec


test/test.hs
module Main where

import Test.Tasty
import Test.Tasty.HUnit

import Data.List
import Data.Ord

main = defaultMain tests

tests :: TestTree
tests = testGroup "Tests" [unitTests]

unitTests = testGroup "Unit tests"
  [ testCase "List comparison (different length)" $
      [1, 2, 3] `compare` [1,2] @?= GT

  -- the following test does not hold
  , testCase "List comparison (same length)" $
      [1, 2, 3] `compare` [1,2,2] @?= LT
  ]


test/Test/AdHoc.hs
module Test.AdHoc (Parser) where

import Data.Void
import Text.Megaparsec

type Parser = Parsec Void String



src/Text/Megaparsec/ANSI/C1.hs
{-# language FlexibleContexts, ScopedTypeVariables #-}
module Text.Megaparsec.ANSI.C1
  (
    -- * C1 escape sequence variations

    escSos
  , escSt

    -- * C1 single 8-bit character variations
  , single8BitSos
  , single8BitSt

  ) where

import Text.Megaparsec
import Data.Proxy
import Data.Semigroup (Semigroup, (<>))

escSos :: forall e s m. (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => m (Tokens s)
escSos = chunk (toChunk 0x1b <> toChunk 0x58) <?> "ESC_SOS"
  where
    toChunk c = tokenToChunk (Proxy :: Proxy s) (toEnum c)
{-# INLINE escSos #-}

escSt :: forall e s m. (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => m (Tokens s)
escSt = chunk (toChunk 0x1b <> toChunk 0x5c) <?> "ESC_ST"
  where
    toChunk c = tokenToChunk (Proxy :: Proxy s) (toEnum c)
{-# INLINE escSt #-}

single8BitSos :: forall e s m. (MonadParsec e s m, Enum (Token s)) => m (Token s)
single8BitSos = single (toEnum 0x98)
{-# INLINE single8BitSos #-}

single8BitSt :: forall e s m. (MonadParsec e s m, Enum (Token s)) => m (Token s)
single8BitSt = single (toEnum 0x9c)
{-# INLINE single8BitSt #-}


src/Text/Megaparsec/ANSI/Common.hs
-- | Common helpers for building terminal parsers.
-- Typically there is no need to import this directly as these helpers are generally used building blocks in this package.
module Text.Megaparsec.ANSI.Common
  (
    -- * Trivial matches
    isC0
  , isSp
  , isDel
  , isEsc
  , isPrint
  , isSpace
  , isControl
  , isGraphic

    -- * ASCII specific
  , isAsciiPrint
  , isAsciiFormatEffector

    -- * ECMA 48 control sequence, ancillary bytes
  , isCsParameterByte
  , isCsIntermediateByte
  , isCsFinalByte

    -- * ISO/IEC 2022 character set designation, ancillary bytes
  , isIso2022IntermediateByte
  , isIso2022StandardFinalByte
  , isIso2022PrivateFinalByte
  ) where

import qualified Data.Char as C

-- | Match the C0 character set. Does not include \SP or \DEL.
isC0 :: (Ord char, Enum char) => char -> Bool
isC0 c = c >= toEnum 0 && c <= toEnum 31
{-# INLINE isC0 #-}

-- | Match the \SP (simply @' '@) character.
isSp :: (Eq char, Enum char) => char -> Bool
isSp c = c == toEnum 32
{-# INLINE isSp #-}

-- | Match the \DEL character.
isDel :: (Eq char, Enum char) => char -> Bool
isDel c = c == toEnum 127
{-# INLINE isDel #-}

-- | Match the \ESC character.
isEsc :: (Eq char, Enum char) => char -> Bool
isEsc c = c == toEnum 0x1b
{-# INLINE isEsc #-}

-- | Match the set of print characters.
isPrint :: (Enum char) => char -> Bool
isPrint = C.isPrint . toEnum . fromEnum
{-# INLINE isPrint #-}

-- | Match the set of whitespace characters.
isSpace :: (Enum char) => char -> Bool
isSpace = C.isSpace . toEnum . fromEnum
{-# INLINE isSpace #-}

-- | Match the set of control characters.
isControl :: (Enum char) => char -> Bool
isControl = C.isControl . toEnum . fromEnum
{-# INLINE isControl #-}

-- | Match the set of graphic characters (including whitespace characters in C0 and \160).
isGraphic :: (Enum char) => char -> Bool
isGraphic = (\c -> C.isPrint c || C.isSpace c) . toEnum . fromEnum
{-# INLINE isGraphic #-}

-- | Match the set of printable ASCII characters.
isAsciiPrint :: (Enum char, Ord char) => char -> Bool
isAsciiPrint c = c >= toEnum 0x20 && c <= toEnum 0x7e
{-# INLINE isAsciiPrint #-}

-- | Match format effector characters FE0 to FE5 (@\b@,@\t@,@\n@,@\v@,@\f@,@\r@)
-- Note that this does not include format effectors in the C1 set:
-- In other words, the following are deliberately excluded:
-- @IND@, @NEL@, @HTS@, @HTJ@, @VTS@, @PLD@, @PLU@, @RI@
isAsciiFormatEffector :: (Enum char, Ord char) => char -> Bool
isAsciiFormatEffector c = c >= toEnum 0x8 && c <= toEnum 0xd
{-# INLINE isAsciiFormatEffector #-}

-- | Control sequence parameter byte; zero or more may follow a CSI.
-- See ECMA-48, section 5.4.
isCsParameterByte :: (Enum char, Ord char) => char -> Bool
isCsParameterByte c = c >= toEnum 0x30 && c <= toEnum 0x3f
{-# INLINE isCsParameterByte #-}

-- | Control sequence intermediate byte; zero or more may follow a CSI and any additional control sequence parameter bytes.
-- See ECMA-48, section 5.4.
isCsIntermediateByte :: (Enum char, Ord char) => char -> Bool
isCsIntermediateByte c = c >= toEnum 0x20 && c <= toEnum 0x2f
{-# INLINE isCsIntermediateByte #-}

-- | Control sequence final byte; a single byte that terminates the control sequence.
-- See ECMA-48, section 5.4.
isCsFinalByte :: (Enum char, Ord char) => char -> Bool
isCsFinalByte c = c >= toEnum 0x40 && c <= toEnum 0x7e
{-# INLINE isCsFinalByte #-}

-- | ISO/IEC-2022 intermediate byte; one or more may follow \ESC to designate a character set.
-- See ECMA-35, section 13.2 and table 5 in particular.
isIso2022IntermediateByte :: (Enum char, Ord char) => char -> Bool
isIso2022IntermediateByte c = c >= toEnum 0x20 && c <= toEnum 0x2f
{-# INLINE isIso2022IntermediateByte #-}

-- | ISO/IEC-2022 standard final byte, subject to registration; a single byte that terminates an ISO/IEC 2022 character designating escape sequence.
-- See ECMA-35, section 13.3
isIso2022StandardFinalByte :: (Enum char, Ord char) => char -> Bool
isIso2022StandardFinalByte c = c >= toEnum 0x40 && c <= toEnum 0x7f
{-# INLINE isIso2022StandardFinalByte #-}

-- | ISO/IEC-2022 final byte reserved for private use; a single byte that terminates an ISO/IEC 2022 character designating escape sequence.
-- See ECMA-35, section 13.3
isIso2022PrivateFinalByte :: (Enum char, Ord char) => char -> Bool
isIso2022PrivateFinalByte c = c >= toEnum 0x30 && c <= toEnum 0x3F
{-# INLINE isIso2022PrivateFinalByte #-}


src/Text/Megaparsec/ANSI/Internal.hs
{-# language ScopedTypeVariables #-}

-- | Internal helpers that normally do not need to be imported.
module Text.Megaparsec.ANSI.Internal
  ( psingleton
  , pappend
  , manyTerminatedBy
  ) where

import Text.Megaparsec
import Data.Proxy (Proxy (..))
import Data.Semigroup (Semigroup, (<>))
import Control.Applicative (liftA2)

-- | Turn a character parser into a string parser
psingleton :: forall e s m. (Stream s, MonadParsec e s m) => m (Token s) -> m (Tokens s)
psingleton = fmap (tokenToChunk (Proxy :: Proxy s))
{-# INLINE psingleton #-}

-- | Append two parse results without requiring @Semigroup (m a)@
pappend :: (MonadParsec e s m, Semigroup a) => m a -> m a -> m a
pappend = liftA2 (<>)
{-# INLINE pappend #-}

-- | Similar to 'manyTill', but includes the terminating chunk (and the parse results are semigroup).
manyTerminatedBy :: (MonadParsec e s m, Semigroup a) => m a -> m a -> m a
manyTerminatedBy p end = go
  where
    go = end <|> pappend p go
{-# INLINE manyTerminatedBy #-}


src/Text/Megaparsec/ANSI/Lexer.hs
{-# language FlexibleContexts, ScopedTypeVariables #-}
-- | Lexer for splitting terminal ("ANSI" / ECMA-48 / ISO 6429) escape sequences from regular text.
-- For many applications this module may be imported by itself.
module Text.Megaparsec.ANSI.Lexer
  (
    -- * Compatibility

    -- $compat
    Single8BitC1Compatibility (..)
  , Iso2022Compatibility (..)

    -- * Regular text

    -- $regtext
  , plainText
  , plainText1
  , printText
  , printText1
  , individualChars
  , individualChars1

    -- * C1 escape sequence variations

    -- $escvar
  , escC1
  , escCsi

    -- * C1 single 8-bit character variations

    -- $singlevar
  , single8BitC1
  , single8BitCsi

    -- * ISO/IEC 2022
    -- $iso2022
  , anyIso2022
  , iso2022CharsetDesignation
  , iso2022AdditionalControlSequence

    -- * ISO/IEC 2022 / C1 single 8-bit character / escape sequence compatible parsers

    -- $c1iso2022compat
  , anyCsi
  , anyAnsiControlSequence
  , anyControlSequence
  , anyControlFunction
  , anyCharacterString
  , anyCommandString
  , anyCommandStringDelimeter
  , anyControlString
  , anyControlStringDelimeter

    -- * Trivial parsers

    -- $trivial
  , esc

  ) where

import Text.Megaparsec
import Text.Megaparsec.ANSI.Internal
import Text.Megaparsec.ANSI.Common
import Text.Megaparsec.ANSI.C1.Lexer
import Data.Semigroup (Semigroup, (<>))
import Data.Proxy

-- $compat

-- | A flag indicating whether C1 control character set should be parsed. Typically this may be provided to a terminal
-- application via a command-line flag.
--
-- Note that VT100 compatible terminals enable/disable the 8-bit C1 character set using escape sequences called
-- @S8C1T@/@S7C1T@. The parsers provided in this package do not attempt to handle stateful parse modes that this would
-- require. However an application can extend the parsers by wrapping the monad transformer provided by megaparsec,
-- 'Text.Megaparsec.ParsecT', around 'State'.
data Single8BitC1Compatibility
  = ExcludeSingle8BitC1 -- ^ Recommended: Do not parse 8-bit C1 control codes (similar to @S7C1T@ in VT100-compatible terminals)
  | IncludeSingle8BitC1 -- ^ Parse 8-bit C1 control codes as well as ESC-prefixed alternatives (similar to @S8C1T@ in VT100-compatible terminals)

-- | A flag indicating whether ISO control characters need to be parsed.
--
-- Note that UTF-8 fits the "other coding system" definition described section 15.4 of ECMA 35. In other words, it exists outside
-- of the ISO 2022 mechanism for shifting character sets.
-- UTF-8 should be prefered when possible since it avoids the various problems inherent to a stateful coding system like ISO 2022.
-- The program @luit@ can be used (and may even be invoked automatically by your terminal emulator) to convert to UTF-8.
-- Use the program @locale@ in order to inspect your environment.
--
-- See also:
-- * https://www.cl.cam.ac.uk/~mgk25/unicode.html#term
-- * https://askubuntu.com/a/133221/17596
-- * https://www.irif.fr/~jch/software/luit
-- * https://invisible-island.net/luit/luit-figures.html
data Iso2022Compatibility
  = ExcludeIso2022 -- ^ Recommended: Do not parse ISO/IEC 2022.
  | IncludeIso2022 -- ^ Parse ISO/IEC 2022

-- $regtext
-- Use these parsers to match sequences characters that have no special meaning in terminals.

-- | Parse consecutive plain text characters (no control characters other than whitespace).
plainText :: (MonadParsec e s m, Stream s, Ord e, Enum (Token s)) => m (Tokens s)
plainText = takeWhileP Nothing isGraphic
{-# INLINE plainText #-}

-- | Parse consecutive plain text characters (no control characters other than whitespace).
-- Fails if it doesn't match at least 1 character.
plainText1 :: (MonadParsec e s m, Stream s, Ord e, Enum (Token s)) => m (Tokens s)
plainText1 = takeWhile1P Nothing isGraphic
{-# INLINE plainText1 #-}

-- | Parse consecutive print characters (no control characters or whitespace other than the normal space character).
printText :: (MonadParsec e s m, Stream s, Ord e, Enum (Token s)) => m (Tokens s)
printText = takeWhileP Nothing isPrint
{-# INLINE printText #-}

-- | Parse consecutive print characters (no control characters or whitespace other than the normal space character).
-- Fails if it doesn't match at least 1 character.
printText1 :: (MonadParsec e s m, Stream s, Ord e, Enum (Token s)) => m (Tokens s)
printText1 = takeWhile1P Nothing isPrint
{-# INLINE printText1 #-}

-- | Parse consecutive individual characters (including C0 and single 8-bit C1 control characters), but not including escape sequences.
-- This can be used in situations where escape sequences are the only control functions of interest.
individualChars :: (MonadParsec e s m, Stream s, Ord e, Enum (Token s), Eq (Token s)) => m (Tokens s)
individualChars = takeWhileP Nothing (not . isEsc)
{-# INLINE individualChars #-}

-- | Parse consecutive individual characters (including C0 and single 8-bit C1 control characters), but not including escape sequences.
-- This can be used in situations where escape sequences are the only control functions of interest.
-- Fails if it doesn't match at least 1 character.
individualChars1 :: (MonadParsec e s m, Stream s, Ord e, Enum (Token s), Eq (Token s)) => m (Tokens s)
individualChars1 = takeWhile1P Nothing (not . isEsc)
{-# INLINE individualChars1 #-}

-- $escvar
-- Control sequences that have a two character ESC-prefixed variation (as well as an 8-bit variation)

-- | Control sequence introducer
-- This parser recognizes only the two character escape-prefixed CSI sequence, not the 8-bit C1 control character.
-- See ECMA-48, section 5.4.
escCsi :: forall e s m. (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => m (Tokens s)
escCsi =
  -- More simply: chunk "\ESC[" :: IsString s => m (Tokens s)
  -- this just elects to use Enum rather than IsString
  chunk (toChunk 0x1b <> toChunk 0x5b) <?> "ESC_CSI"
  where
    toChunk c = tokenToChunk (Proxy :: Proxy s) (toEnum c)
{-# INLINE escCsi #-}

-- | Escape sequence equivalents to the C1 character set
escC1 :: forall e s m. (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => m (Tokens s)
escC1 = (psingleton esc `pappend` psingleton (satisfy (\c -> c >= toEnum 0x40 && c <= toEnum 0x5F))) <?> "ESC_C1"
{-# INLINE escC1 #-}

-- $singlevar
-- Note that the ECMA-48 standard allows for certain ESC-prefixed control sequences to be substituted by a single 8-bit
-- C1 control code character (in 8-bit environments). However, according to
-- <https://en.wikipedia.org/wiki/ANSI_escape_code#Escape_sequences Wikipedia>, the 2-byte escape sequences are
-- commonly used in modern environments due to incompatible interpretations in some UTF-8 and CP-1252 character sets.
--
-- For that reason it is recommended to use ESC-prefixed codes. However, in some rare instances where this is not possible
-- the compatible parsers in section following this one can be used. Individual c1 character parsers are provided here for
-- the sake of completeness only.

-- | Single character, C1 character set
single8BitC1 :: (MonadParsec e s m, Ord (Token s), Enum (Token s)) => m (Token s)
single8BitC1 = satisfy (\c -> c >= toEnum 0x80 && c <= toEnum 0x9F) <?> "C1"
{-# INLINE single8BitC1 #-}

-- | Single character, control sequence introducer
single8BitCsi :: (MonadParsec e s m, Enum (Token s)) => m (Token s)
single8BitCsi = single (toEnum 0x9b)
{-# INLINE single8BitCsi #-}

-- | Match any ISO/IEC 2022 control function.
-- Note that this includes some control functions in C0 and C1.
-- 'iso2022AdditionalControlSequence' and 'iso2022CharsetDesignation' match codes that are unique to ISO/IEC 2022.
anyIso2022 :: forall e s m. (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => m (Tokens s)
anyIso2022 =
  (psingleton (oneOf (map toEnum [
    -- Included in C0
    toEnum 0x0e, toEnum 0x0f,
    -- Included in C1 8-bit codes
    toEnum 0x8e, toEnum 0x8f
  ]))
  <|>
    -- Included in C1 escape sequneces
    chunk (toChunk 0x1b <> toChunk 0x4e) <|> chunk (toChunk 0x1b <> toChunk 0x4f)
  <|>
    -- GL/GR designations not present in C1 (anything larger than \ESC\x5F)
    iso2022AdditionalControlSequence
  <|>
    -- Character set designations for the working character sets (C0,C1,G0,G1,G2,G3,etc)
    iso2022CharsetDesignation
  ) <?> "ISO2022"
  where
    toChunk c = tokenToChunk (Proxy :: Proxy s) (toEnum c)

-- | Matches ISO/IEC 2022 control sequences that designate a character set to use.
-- See 'anyIso2022' to see how this fits in with the other codes found in ISO/IEC 2022.
-- See ECMA 35, section 13.2.2.
iso2022CharsetDesignation :: (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => m (Tokens s)
iso2022CharsetDesignation =
  psingleton esc `pappend` intermediateBytes `pappend` finalByte
  where
    intermediateBytes = takeWhile1P (Just "ISO/IEC 2022 intermediate byte") isIso2022IntermediateByte
    finalByte = psingleton (satisfy (\c -> isIso2022StandardFinalByte c || isIso2022PrivateFinalByte c))

-- | Additional control functions (escape sequences) unique to ISO/IEC 2022.
-- These are not present in the C0, C1 sets, or any other ANSI escape sequences reserved by ECMA-48.
-- See 'anyIso2022' to see how this fits in with the other codes found in ISO/IEC 2022.
iso2022AdditionalControlSequence :: forall e s m. (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => m (Tokens s)
iso2022AdditionalControlSequence =
  psingleton esc `pappend` psingleton (satisfy (\c -> c >= toEnum 0x60 && c <= toEnum 0x7E))
{-# INLINE iso2022AdditionalControlSequence #-}

-- $c1compat
-- Use these parsers when support for single-byte C1 characters is a requirement.

-- | Single character or escape sequence, control sequence introducer.
anyCsi :: (MonadParsec e s m, Semigroup (Tokens s), Enum (Token s)) => Single8BitC1Compatibility -> m (Tokens s)
anyCsi ExcludeSingle8BitC1 = escCsi <?> "CSI"
anyCsi IncludeSingle8BitC1 = (escCsi <|> psingleton single8BitCsi) <?> "CSI"
{-# INLINE anyCsi #-}

-- | Any CSI-prefixed control sequence.
anyAnsiControlSequence :: (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => Single8BitC1Compatibility -> m (Tokens s)
anyAnsiControlSequence s8c1Compat = anyCsi s8c1Compat `pappend` parameterBytes `pappend` intermediateBytes `pappend` finalByte
  where
    parameterBytes = takeWhileP (Just "CSI parameter byte") isCsParameterByte
    intermediateBytes = takeWhileP (Just "CSI intermediate byte") isCsIntermediateByte
    finalByte = psingleton (satisfy isCsFinalByte)
{-# INLINE anyAnsiControlSequence #-}

-- | Match any control sequence.
-- Uses lookahead to quickly fail on any @ESC@ (or @CSI@) character prefix.
anyControlSequence :: (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => Single8BitC1Compatibility -> Iso2022Compatibility -> m (Tokens s)
anyControlSequence s8c1Compat iso2022Compat =
  let
    matchPrefix = case s8c1Compat of
      IncludeSingle8BitC1 -> esc <|> single8BitCsi
      ExcludeSingle8BitC1 -> esc
    matchEcma48 = anyAnsiControlSequence s8c1Compat <|> escC1
    matchIso2022 = case iso2022Compat of
      IncludeIso2022 -> iso2022CharsetDesignation <|> iso2022AdditionalControlSequence
      ExcludeIso2022 -> empty
  in
    lookAhead matchPrefix *> (matchEcma48 <|> matchIso2022)
{-# INLINE anyControlSequence #-}

-- | Any terminal control character or sequence of characters.
-- Differs from 'Text.Megaparsec.controlChar' in that multi-byte sequences are fully consumed as lexemes.
anyControlFunction :: (MonadParsec e s m, Stream s, Semigroup (Tokens s), Enum (Token s)) => Single8BitC1Compatibility -> Iso2022Compatibility -> m (Tokens s)
anyControlFunction s8c1Compat iso2022Compat =
  (anyControlSequence s8c1Compat iso2022Compat <|> psingleton controlChar)
  where
    controlChar = satisfy isControl <?> "control character"
{-# INLINE anyControlFunction #-}

-- | Any control string.
-- See ECMA-48, section 5.6.
anyControlString :: (MonadParsec e s m, Semigroup (Tokens s), Enum (Token s)) => Single8BitC1Compatibility -> m (Tokens s)
anyControlString s8c1Compat =
  anyCharacterString s8c1Compat <|> anyCommandString s8c1Compat
{-# INLINE anyControlString #-}

-- | Any control string delimeter (DCS, SOS, OSC, PM, APC).
anyControlStringDelimeter :: (MonadParsec e s m, Semigroup (Tokens s), Enum (Token s)) => Single8BitC1Compatibility -> m (Tokens s)
anyControlStringDelimeter ExcludeSingle8BitC1 =
  psingleton esc `pappend` psingleton (satisfy (\c -> (c >= toEnum 0x5d && c <= toEnum 0x5f) || c == toEnum 0x50 || c == toEnum 0x58))
anyControlStringDelimeter IncludeSingle8BitC1 =
  anyControlStringDelimeter ExcludeSingle8BitC1 <|> psingleton (satisfy (\c -> (c >= toEnum 0x9d && c <= toEnum 0x9f) || c == toEnum 0x90 || c == toEnum 0x98))
{-# INLINE anyControlStringDelimeter #-}

-- | Any character string (a type of control string).
-- Appears to be recognized, but ignored, by a tested terminal emulator.
--
-- E.g.
--
-- >>> putStrLn "abc\ESCXtest\ESC\\def"
-- abcdef
--
anyCharacterString :: (MonadParsec e s m, Semigroup (Tokens s), Enum (Token s)) => Single8BitC1Compatibility -> m (Tokens s)
anyCharacterString s8c1Compat =
  sos `pappend` manyTerminatedBy (psingleton anySingle) st
  where
    (sos, st) = case s8c1Compat of
      ExcludeSingle8BitC1 ->
        (escSos, escSt)
      IncludeSingle8BitC1 ->
        (escSos <|> psingleton single8BitSos, escSt <|> psingleton single8BitSt)
{-# INLINE anyCharacterString #-}

-- | Any command string delimeter @DCS@, @OSC@, @PM@, @APC@.
anyCommandStringDelimeter :: (MonadParsec e s m, Enum (Token s), Semigroup (Tokens s)) => Single8BitC1Compatibility -> m (Tokens s)
anyCommandStringDelimeter ExcludeSingle8BitC1 =
  psingleton esc `pappend` psingleton (satisfy (\c -> (c >= toEnum 0x5d && c <= toEnum 0x5f) || c == toEnum 0x50))
anyCommandStringDelimeter IncludeSingle8BitC1 =
  anyCommandStringDelimeter ExcludeSingle8BitC1 <|> psingleton (satisfy (\c -> (c >= toEnum 0x9d && c <= toEnum 0x9f) || c == toEnum 0x90))
{-# INLINE anyCommandStringDelimeter #-}

-- | Any command string (a type of control string).
-- * @DCS@ appears to be recognized, but ignored, by a tested terminal emulator.
-- * @PM@ appears to be unrecognized by a tested terminal emulator.
-- * @OSC@ appears to be unrecognized by a tested terminal emulator.
-- * @APC@ appears to be unrecognized by a tested terminal emulator.
--
-- E.g.
--
-- >>> putStrLn "abc\ESCPtest\ESC\\def"
-- abcdef
-- >>> putStrLn "abc\ESC^test\ESC\\def"
-- abc␛^test␛\def
-- >>> putStrLn "abc\ESC]test\ESC\\def"
-- abc␛]test␛\def
-- >>> putStrLn "abc\ESC_test\ESC\\def"
-- abc␛_test␛\def
--
anyCommandString :: (MonadParsec e s m, Semigroup (Tokens s), Enum (Token s)) => Single8BitC1Compatibility -> m (Tokens s)
anyCommandString s8c1Compat =
  anyCommandStringDelimeter s8c1Compat `pappend` takeWhileP (Just "command string byte") isCommandStringByte `pappend` st
  where
    -- We assume that neither UTF-8, nor C1 format effectors are ot accepted since ECMA-48 specifies the ranges explicitly.
    isCommandStringByte c = isAsciiPrint c || isAsciiFormatEffector c
    st = case s8c1Compat of
      ExcludeSingle8BitC1 -> escSt
      IncludeSingle8BitC1 -> escSt <|> psingleton single8BitSt
{-# INLINE anyCommandString #-}

-- $trivial

-- | Escape character, commonly prefixed to escape sequences.
esc :: (MonadParsec e s m, Enum (Token s)) => m (Token s)
esc = single (toEnum 0x1b)
{-# INLINE esc #-}

